[stable/elasticsearch] Add user/password option for basic auth on client ingress

donbowman opened the pull request:
If `client.ingress.user` and `client.ingress.password` are set,
and client ingress is enabled, HTTP basic authentication will be
enabled on the client ingress.

This allows for (modest) securing of external access to the system.

Signed-off-by: Don Bowman <db@donbowman.ca>

<!--
Thank you for contributing to helm/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/helm/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/helm/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a CircleCI
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->

#### What this PR does / why we need it:

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #

#### Special notes for your reviewer:

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x ] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x ] Chart Version bumped
- [ x] Variables are documented in the README.md

donbowman committed:
Add user/password option for basic auth on client ingress

If `client.ingress.user` and `client.ingress.password` are set,
and client ingress is enabled, HTTP basic authentication will be
enabled on the client ingress.

This allows for (modest) securing of external access to the system.

Signed-off-by: Don Bowman <db@donbowman.ca>

desaintmartin commented on stable/elasticsearch/README.md:
```diff
@@ -92,6 +92,8 @@ The following table lists the configurable parameters of the elasticsearch chart
 | `client.initResources`               | Client initContainer resources requests & limits                    | `{}`                                                |
 | `client.additionalJavaOpts`          | Parameters to be added to `ES_JAVA_OPTS` environment variable for client | `""`                                           |
 | `client.ingress.enabled`             | Enable Client Ingress                                               | `false`                                             |
+| `client.ingress.user`                | If this & password are set, enable basic-auth on ingress            | `nil`                                               |
```
Actually, this is activated even if password is not set and only user is set. Is that intended?

donbowman replied:
If only the user is set, the password is blank. I guess I could check that case, but then a blank password is disallowed.

So yes, its what i intended, but am not really sure if there's value in it. Do you want me to add an if user and password check to both? I think it works as-is

```
--- /dev/null
+++ b/stable/elasticsearch/templates/client-auth.yaml
@@ -0,0 +1,11 @@
+{{- if .Values.client.ingress.user -}}
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: '{{ include "elasticsearch.client.fullname" . }}-auth'
+type: Opaque
+data:
+  auth: {{ printf "%s:{PLAIN}%s\n" .Values.client.ingress.user .Values.client.ingress.password | b64enc | quote }}
+{{- end }}
+
diff --git a/stable/elasticsearch/templates/client-ingress.yaml b/stable/elasticsearch/templates/client-ingress.yaml
index 0564589f..8abc2ce6 100644
--- a/stable/elasticsearch/templates/client-ingress.yaml
+++ b/stable/elasticsearch/templates/client-ingress.yaml
@@ -15,6 +15,11 @@ metadata:
   annotations:
 {{ toYaml . | indent 4 }}
 {{- end }}
+{{- if .Values.client.ingress.user -}}
+    nginx.ingress.kubernetes.io/auth-type: basic
+    nginx.ingress.kubernetes.io/auth-secret: '{{ include "elasticsearch.client.fullname" . }}-auth'
+    nginx.ingress.kubernetes.io/auth-realm: "Authentication-Required"
+{{- end }}
```

donbowman committed:
Set basic authentication only if both user & password are set

Only create secret and annotation if both client.ingress.user
and client.ingress.password are set.

Add commented-out config in values.yaml for client.ingress.user
and client.ingress.password.

Signed-off-by: Don Bowman <db@donbowman.ca>
